### ISSUE TITLE
opt-in files on an individual basis using comment in file

### ISSUE THREAD

**Issue by rsyring** [astral-sh/ruff#6156]
On legacy code bases its sometimes not possible to get buy-in to do a project wide reformat.  However, incremental migration using ruff, especially on brand new files or code that is unlikely to experience merge conflicts with other PRs due to reformating, is easier to justify and implement.  

Right now, that means updating the `include` setting for every file that should be handled.  In a large project, that list is going to get really big and dominate the config file.  I thought it would be better to have a special comment, maybe `# include-qa: ruff`, along with a setting like `include-by-comment: true`, that would instruct ruff to include those files.  When it's false, `# include-qa: ruff` would be flagged as a fixable linting error.  That way, when a codebase has been migrated completely, modify that setting to false, use `--fix` and all the `include-qa` comments can be removed by ruff.

I suggest `# include-qa: ruff` in the hope that other tools might be persuaded to use the same system.  E.g. `# include-qa: ruff black`

**Comment by charliermarsh** [comment]
What about adding `# ruff: noqa` to each file, to exempt them all? We could then support flagging unused `# ruff: noqa` comments, and automatically remove them as you gained compatibility.

**Comment by rsyring** [comment]
So like `--add-noqa` but at the file instead of line level.  It's better than the current options, so if that's where we end up, I'll take it.  Given its only a single line change its unlikely to create merge conflicts.  I like automatically removing those lines when they are no longer needed.

However, that still entails editing every Python file in the repository.  Given the context I'm considering with this request, large code bases hesitant to enable wholesale changes, it could still create more friction than an opt-in solution.

On the other hand, its likely that linting and formatting would be desired and, at least in black, the only mechanism I've found to accomplish these same goals is to add `# fmt: off` at the top of every file.  If adding a single line is necessary, adding two lines isn't much harder.

**Comment by zanieb** [comment]
Related https://github.com/astral-sh/ruff/issues/1149

**Comment by rsyring** [comment]
FWIW: I ended up creating a Python script to run black and ruff only if a sentinel value was present in the file.  This does solve the problem when running the tools from the terminal but has enabled lint/format on save functionality, which is the primary use case right now.  In VS Code, this even shows linting in the editor due to its support for "problem matching."

```python
#!/usr/bin/env python
"""
    Can be used to support format-on-save functionality in code editors on a per-file basis.

    Configure your editor to run this file on save.  VS Code example:

    tasks.json:
    {
        // See https://go.microsoft.com/fwlink/?LinkId=733558
        // for the documentation about the tasks.json format
        "version": "2.0.0",
        "tasks": [
            {
                "label": "code-qa",
                "type": "shell",
                "command": "./bin/code-qa.py ${file}",
                "presentation": {
                    // Comment out to make terminal reveal every time task runs
                    "reveal": "never"
                },
                "problemMatcher": {
                    "owner": "py",
                    "fileLocation": ["relative", "${workspaceFolder}"],
                    "pattern": {
                    "regexp": "^(.*):(\\d+):(\\d+):\\s+([^\\s]+)\\s+(.*)$",
                    "file": 1,
                    "line": 2,
                    "column": 3,
                    "code": 4,
                    "message": 5
                    }
                }
            }
        ]
    }

    user or workspace settings.json:

    "runOnSave.commands": [
        {
            "match": ".*\\.py$",
            "command": "workbench.action.tasks.runTask",
            "args": "code-qa",
            "runIn": "vscode",
        }
    ]

    Above requires https://marketplace.visualstudio.com/items?itemName=pucelle.run-on-save
"""

import subprocess
import sys


def run(*args):
    return subprocess.run(args)


source_fpath = sys.argv[1]
result: subprocess.CompletedProcess = run('rg', '-q', '# code-qa: on', source_fpath)
if result.returncode == 0:

    run('ruff', 'check', '-q', '--fix', source_fpath)

    run('black', '--fast', '-q', source_fpath)
```
